File: frontend/lib/models/sofa_loan.dart
class SofaLoan {
  final int id;
  final int groupId;
  final String name;
  final double principalAmount;
  final String disbursedDate;
  final String status; // 'active' | 'closed'
  final String? closedDate;
  final double outstanding;
  final DateTime createdAt;
  final DateTime updatedAt;

  const SofaLoan({
    required this.id,
    required this.groupId,
    required this.name,
    required this.principalAmount,
    required this.disbursedDate,
    required this.status,
    this.closedDate,
    required this.outstanding,
    required this.createdAt,
    required this.updatedAt,
  });

  bool get isActive => status == 'active';

  factory SofaLoan.fromJson(Map<String, dynamic> json) {
    final now = DateTime.now();
    return SofaLoan(
      id: json['id'] as int,
      groupId: json['group_id'] as int,
      name: json['name'] as String,
      principalAmount: (json['principal_amount'] as num).toDouble(),
      disbursedDate: json['disbursed_date'] as String,
      status: json['status'] as String,
      closedDate: json['closed_date'] as String?,
      outstanding: (json['outstanding'] as num).toDouble(),
      createdAt: DateTime.tryParse(json['created_at'] as String? ?? '') ?? now,
      updatedAt: DateTime.tryParse(json['updated_at'] as String? ?? '') ?? now,
    );
  }
}

File: backend/app/schemas/sofa_loan.py
from datetime import date, datetime

from pydantic import BaseModel, Field


class SofaLoanCreate(BaseModel):
    principal_amount: float = Field(..., gt=0)
    disbursed_date: date


class SofaLoanRead(BaseModel):
    id: int
    group_id: int
    name: str
    principal_amount: float
    disbursed_date: date
    status: str
    closed_date: date | None = None
    outstanding: float = 0.0
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
